ujt/generate_data.py

```diff
@@ -0,0 +1,154 @@
+
+from . import utils
+
+# define the mock data in a simpler format before generating protobufs
+# service and endpoint names correspond 1:1
+SERVICE_NAMES = [
+    "APIServer",
+    "WebServer",
+    "GameService",
+    "LeaderboardService",
+    "ProfileService",
+    "StoreService",
+    "GameDB",
+    "ProfileDB",
+    "ExternalAuthProvider",
+    "ExternalPaymentProvider",
+]
+SERVICE_ENDPOINT_NAME_MAP = {
+    "APIServer": ["StartGame", "UpdateGameState"],
+    "WebServer": ["ViewProfile", "ViewLeaderboard"],
+    "GameService": ["GetPlayerLocation", "GetScore"],
+    "LeaderboardService": [
+        "GetLeaderboard", "SetUserHighScore", "GetUserHighScore"
+    ],
+    "ProfileService": ["Authenticate", "GetUserInfo"],
+    "StoreService": ["BuyCurrency"],
+    "GameDB": ["ReadHighScore", "WriteHighScore"],
+    "ProfileDB": ["ReadFriendsList", "WriteFriendsList"],
+    "ExternalAuthProvider": [],
+    "ExternalPaymentProvider": [],
+}
+# define a dependency map from endpoint to its dependencies
+# each tuple represents a dependency in the form (target_service_name, target_endpoint_name)
+# note we use empty string instead of None to match the protobuf convention for "unset" fields
+ENDPOINT_DEPENDENCY_MAP: DefaultDict[str, List[Tuple[str,
+                                                     str]]] = defaultdict(list)
+ENDPOINT_DEPENDENCY_MAP.update({
+    # APIServer
+    "StartGame": [("APIServer", "UpdateGameState")],
+    "UpdateGameState": [
+        ("GameService", "GetPlayerLocation"),
+        ("GameService", "GetScore"),
+        ("LeaderboardService", "SetUserHighScore"),
+    ],
+    # WebServer
+    "ViewLeaderboard": [("LeaderboardService", "GetLeaderboard")],
+    "ViewProfile": [
+        ("ProfileService", "Authenticate"),
+        ("ProfileService", "GetUserInfo"),
+    ],
+    "BuyCurrency": [("StoreService", "VerifyPayment")],
+    # LeaderboardService
+    "GetLeaderboard": [("GameDB", "ReadHighScore")],
+    "SetUserHighScore": [("GameDB", "WriteHighScore")],
+    "GetUserHighScore": [("GameDB", "ReadHighScore")],
+    # ProfileService
+    "Authenticate": [("ExternalAuthProvider", "")],
+    "GetUserInfo": [("LeaderboardService", "GetUserHighScore"),
+                    ("ProfileDB", "ReadFriendsList")],
+    # StoreService
+    "VerifyPayment": [("ExternalPaymentProvider", "")],
+})
+# client names and user journeys correspond 1:1
+CLIENT_NAMES = ["MobileClient", "WebBrowser"]
+CLIENT_USER_JOURNEY_NAME_MAP = {
+    "MobileClient": ["Play a Game"],
+    "WebBrowser": [
+        "View Leaderboard",
+        "View Profile",
+        "Conduct Microtransaction",
+    ],
+}
+USER_JOURNEY_DEPENDENCY_MAP = {
+    # MobileClient
+    "Play a Game": [("APIServer", "StartGame")],
+    # WebClient
+    "View Leaderboard": [("WebServer", "ViewLeaderboard")],
+    "View Profile": [("WebServer", "ViewProfile")],
+    "Conduct Microtransaction": [("WebServer", "BuyCurrency")],
+}
+
+
+def generate_services():
```

Comment:
`generate_services()` and `generate_clients()` have identical structure. The former generates services (containing endpoints, which contain dependencies), while the latter generates clients (containing user journeys, which contain dependencies). We can think of services and clients as graph nodes, and endpoints and user journeys as "subnodes".

Should we refactor these into a single method `generate_nodes(node_type, subnode_type)` and call it with `generate_nodes(Service, Endpoint)` and `generate_nodes(Client, UserJourney)`? Or is it clearer to leave it as two methods like this?

My intuition is to refactor them into one method, as it can prevent errors where we changed one method but didn't change the other. However, it not be immediately obvious unclear to an outside reader.

Reply:
Are we confident that clients and services will not diverge in the future? Since we think of them as different types of nodes, it seems more likely that they might be instantiated differently.